File: php/src/io/mxCellCodec.php
<?php
/**
 * Copyright (c) 2006-2013, Gaudenz Alder
 */
class mxCellCodec extends mxObjectCodec
{

	/**
	 * Class: mxCellCodec
	 *
	 * Codec for <mxCell>s. This class is created and registered
	 * dynamically at load time and used implicitely via <mxCodec>
	 * and the <mxCodecRegistry>.
	 *
	 * Transient Fields:
	 *
	 * - children
	 * - edges
	 * - states
	 * - overlay
	 * - mxTransient
	 *
	 * Reference Fields:
	 *
	 * - parent
	 * - source
	 * - target
	 * 
	 * Constructor: mxObjectCodec
	 *
	 * Constructs a new codec for the specified template object.
	 * The variables in the optional exclude array are ignored by
	 * the codec. Variables in the optional idrefs array are
	 * turned into references in the XML. The optional mapping
	 * may be used to map from variable names to XML attributes.
	 *
	 * Parameters:
	 *
	 * template - Prototypical instance of the object to be
	 * encoded/decoded.
	 * exclude - Optional array of fieldnames to be ignored.
	 * idrefs - Optional array of fieldnames to be converted to/from
	 * references.
	 * mapping - Optional mapping from field- to attributenames.
	 */
	function mxCellCodec($template)
	{
		parent::mxObjectCodec($template, array("children", "edges", "states",
			"overlay", "mxTransient"), array("parent",
			"source", "target"));
	}
	
	/**
	 * Override <mxObjectCodec.isExcluded>.
	 */
	function isExcluded($obj, $attr, $value, $isWrite)
	{
		return parent::isExcluded($obj, $attr, $value, $isWrite) ||
				($isWrite && $attr == "value" && is_object($value) &&
			   get_class($value) == "DOMElement");
	}

	/**
	 * Override <mxObjectCodec.afterEncode>.
	 */
	function afterEncode($enc, $obj, $node)
	{
		if (is_object($obj->value) && get_class($obj->value) == "DOMElement")
		{
			// Wraps the graphical annotation up in the
			// user object (inversion) by putting the
			// result of the default encoding into
			// a clone of the user object (node type 1)
			// and returning this cloned user object.
			$tmp = $node;

			$node = $enc->document->importNode($obj->value, true);
			$node->appendChild($tmp);
						
			// Moves the id attribute to the outermost
			// XML node, namely the node which denotes
			// the object boundaries in the file.
			$id = $tmp->getAttribute("id");
			$node->setAttribute("id", $id);
			$tmp->removeAttribute("id");
		}

		return $node;
	}

	/**
	 * Override <mxObjectCodec.beforeDecode>.
	 */
	function beforeDecode($dec, $node, &$obj)
	{
		$inner = $node;
		$classname = $this->getName();
		
		if ($node->nodeName != $classname)
		{
			// Passes the inner graphical annotation node to the
			// object codec for further processing of the cell.
			$tmp = $node->getElementsByTagName($classname)->item(0);
			
			if (isset($tmp) && $tmp->parentNode == $node)
			{
				$inner = $tmp;

				// Removes annotation and whitespace from node
				$tmp2 = $tmp->previousSibling;

				while (isset($tmp2) && $tmp2->nodeType == XML_TEXT_NODE)
				{
					$tmp3 = $tmp2->previousSibling;

					if (strlen(trim($tmp2->textContent)) == 0)
					{
						$tmp2->parentNode->removeChild($tmp2);
					}
					
					$tmp2 = $tmp3;
				}
				
				// Removes more whitespace
				$tmp2 = $tmp->nextSibling;
				
				while (isset($tmp2) && $tmp2->nodeType == XML_TEXT_NODE)
				{
					$tmp3 = $tmp2->previousSibling;
					
					if (strlen(trim($tmp2->textContent)) == 0)
					{
						$tmp2->parentNode->removeChild($tmp2);
					}
					
					$tmp2 = $tmp3;
				}
				
				$tmp->parentNode->removeChild($tmp);
			}
			else
			{
				$inner = null;
			}
			
			// Creates the user object out of the XML node
			$obj->value = $node->cloneNode(true);
			$id = $obj->value->getAttribute("id");
			
			if (strlen($id) > 0)
			{
				$obj->setId($id);
				$obj->value->removeAttribute("id");
			}
		}
		else
		{
			$obj->setId($node->getAttribute("id"));
		}
			
		// Preprocesses and removes all Id-references
		// in order to use the correct encoder (this)
		// for the known references to cells (all).
		if (isset($inner))
		{
			for ($i = 0; $i < sizeof($this->idrefs); $i++)
			{
				$attr = $this->idrefs[$i];
				$ref = $inner->getAttribute($attr);
				
				if (strlen($ref) > 0)
				{
					$inner->removeAttribute($attr);
					$object = (isset($dec->objects[$ref])) ? $dec->objects[$ref] : null;
					
					if (!isset($object))
					{
						$object = $dec->lookup($ref);
					}
					
					if (!isset($object))
					{
						// Needs to decode forward reference
						$element = $dec->getElementById($ref);

						if (isset($element))
						{
							$decoder = mxCodecRegistry::$codecs[$element->nodeName];
	
							if (!isset($decoder))
							{
								$decoder = $this;
							}

							$object = $decoder->decode($dec, $element);
						}
					}
											
					$obj->$attr = $object;
				}
			}
		}
		
		return $inner;
	}

}

mxCodecRegistry::register(new mxCellCodec(new mxCell()));
?>

File: dotnet/src/io/mxCellCodec.cs
// Copyright (c) 2007-2008, Gaudenz Alder
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace com.mxgraph
{
    /// <summary>
    /// Codec for mxCells. This class is created and registered
    /// dynamically at load time and used implicitely via mxCodec
    /// and the mxCodecRegistry.
    /// </summary>
    public class mxCellCodec : mxObjectCodec
    {
        /// <summary>
        /// Constructs a new cell codec.
        /// </summary>
        public mxCellCodec()
            : this(new mxCell(), new string[] { "children", "edges" }, new string[] { "parent", "source", "target" },
                    null) { }

        /// <summary>
        /// Constructs a new cell codec for the given template.
        /// </summary>
        public mxCellCodec(Object template) : this(template, null, null, null) { }

        /// <summary>
        /// Constructs a new cell codec for the given arguments.
        /// </summary>
        public mxCellCodec(Object template, string[] exclude, string[] idrefs,
                Dictionary<string, string> mapping)
            : base(template, exclude, idrefs, mapping) { }

        /// <summary>
        /// Excludes user objects that are XML nodes.
        /// </summary>
        public override bool IsExcluded(Object obj, string attr, Object value,
                bool isWrite)
        {
            return exclude.Contains(attr)
                    || (isWrite && attr.Equals("value") && value is XmlNode && ((XmlNode)value)
                            .NodeType == XmlNodeType.Element);
        }

        /// <summary>
        /// Encodes an mxCell and wraps the XML up inside the
        /// XML of the user object (inversion).
        /// </summary>
        public override XmlNode AfterEncode(mxCodec enc, Object obj, XmlNode node)
        {
            if (obj is mxCell && node is XmlElement)
            {
                mxCell cell = (mxCell)obj;

                if (cell.Value != null)
                {
                    if (cell.Value is XmlNode)
                    {
                        // Wraps the graphical annotation up in the
                        // user object (inversion) by putting the
                        // result of the default encoding into
                        // a clone of the user object (node type 1)
                        // and returning this cloned user object.
                        XmlElement tmp = (XmlElement)node;
                        node = enc.Document.ImportNode((XmlNode)cell.Value, true);
                        node.AppendChild(tmp);

                        // Moves the id attribute to the outermost
                        // XML node, namely the node which denotes
                        // the object boundaries in the file.
                        String id = tmp.GetAttribute("id");
                        ((XmlElement)node).SetAttribute("id", id);
                        tmp.RemoveAttribute("id");
                    }
                }
            }

            return node;
        }

        /// <summary>
        /// Decodes an mxCell and uses the enclosing XML node as
        /// the user object for the cell (inversion).
        /// </summary>
        public override XmlNode BeforeDecode(mxCodec dec, XmlNode node, Object obj)
        {
            XmlElement inner = (XmlElement)node;

            if (obj is mxCell)
            {
                mxCell cell = (mxCell)obj;
                String classname = GetName();

                if (!node.Name.Equals(classname))
                {
                    // Passes the inner graphical annotation node to the
                    // object codec for further processing of the cell.
                    XmlNode tmp = inner.GetElementsByTagName(classname)[0];

                    if (tmp != null && tmp.ParentNode == node)
                    {
                        inner = (XmlElement)tmp;

                        // Removes annotation and whitespace from node
                        XmlNode tmp2 = tmp.PreviousSibling;

                        while (tmp2 != null && tmp2.NodeType == XmlNodeType.Text)
                        {
                            XmlNode tmp3 = tmp2.PreviousSibling;

                            if (tmp2.Value.Trim().Length == 0)
                            {
                                tmp2.ParentNode.RemoveChild(tmp2);
                            }

                            tmp2 = tmp3;
                        }

                        // Removes more whitespace
                        tmp2 = tmp.NextSibling;

                        while (tmp2 != null && tmp2.NodeType == XmlNodeType.Text)
                        {
                            XmlNode tmp3 = tmp2.PreviousSibling;

                            if (tmp2.Value.Trim().Length == 0)
                            {
                                tmp2.ParentNode.RemoveChild(tmp2);
                            }

                            tmp2 = tmp3;
                        }

                        tmp.ParentNode.RemoveChild(tmp);
                    }
                    else
                    {
                        inner = null;
                    }

                    // Creates the user object out of the XML node
                    XmlElement value = (XmlElement)node.CloneNode(true);
                    cell.Value = value;
                    String id = value.GetAttribute("id");

                    if (id != null)
                    {
                        cell.Id = id;
                        value.RemoveAttribute("id");
                    }
                }
                else
                {
                    cell.Id = ((XmlElement)node).GetAttribute("id");
                }

                // Preprocesses and removes all Id-references
                // in order to use the correct encoder (this)
                // for the known references to cells (all).
                if (inner != null && idrefs != null)
                {
                    foreach (string attr in idrefs)
                    {
                        string rf = inner.GetAttribute(attr);

                        if (rf != null && rf.Length > 0)
                        {
                            inner.RemoveAttribute(attr);
                            Object tmp = (dec.Objects.ContainsKey(rf)) ? dec.Objects[rf] : null;

                            if (tmp == null)
                            {
                                tmp = dec.Lookup(rf);
                            }

                            if (tmp == null)
                            {
                                // Needs to decode forward reference
                                XmlNode element = dec.GetElementById(rf);

                                if (element != null)
                                {
                                    mxObjectCodec decoder = mxCodecRegistry
                                            .GetCodec(element.Name);

                                    if (decoder == null)
                                    {
                                        decoder = this;
                                    }

                                    tmp = decoder.Decode(dec, element);
                                }
                            }

                            SetFieldValue(obj, attr, tmp);
                        }
                    }
                }
            }

            return inner;
        }

    }

}
